### PR TITLE
[quant][be] Add a check in prepare_qat to make sure the model is in training mode

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70009
* #69959
* __->__ #69879

Summary:
att

Test Plan:
```
python test/test_quantization.py TestQuantizationAwareTraining
```
Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33080989](https://our.internmc.facebook.com/intern/diff/D33080989)

### CODE REVIEW

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

File: torch/ao/quantization/quantize.py [hunk 1]
```diff
@@ -435,6 +435,7 @@ def prepare_qat(model, mapping=None, inplace=False, allow_list=None):
         allow_list: a set that lists out allowable modules to be propagated with qconfig
     """
     torch._C._log_api_usage_once("quantization_api.quantize.prepare_qat")
+    assert model.training, "prepare_qat only works on models in training mode"
```
Comment (supriyar) [inline torch/ao/quantization/quantize.py:438]: Does it make sense to not have this restriction on the model? our APIs should work on the model irrespective of it being in `train`/`eval` mode.

Comment (jerryzh168) [reply torch/ao/quantization/quantize.py:438]: we do have this assumption currently, although it's implicit. I think we can relax the constraint later

Comment (jerryzh168) [reply torch/ao/quantization/quantize.py:438]: we will need to refactor these to remove the constraint: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fuser_method_mappings.py#L32, will add the PR in this stack

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

**Review by vkuzo — APPROVED** [review]

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).

**Comment by jerryzh168** [comment]
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33080989).
